src/pybind/mgr/dashboard/controllers/osd.py

```diff
@@ -63,17 +63,43 @@ def add_id(osd):
         }
         return resp if svc_id is None else resp[int(svc_id)]
 
+    @staticmethod
+    def _get_smart_data(svc_id):
+        # type: (str) -> dict
+        """
+        Returns S.M.A.R.T data for the given OSD ID.
+        :type svc_id: Numeric ID of the OSD
+        """
+        devices = CephService.send_command(
+            'mon', 'device ls-by-daemon', who='osd.{}'.format(svc_id))
+        smart_data = {}
+        for dev_id in [d['devid'] for d in devices]:
+            if dev_id not in smart_data:
+                dev_smart_data = mgr.remote('devicehealth', 'do_scrape_daemon', 'osd', svc_id,
+                                            dev_id)
```

Comment:
Any idea/estimation of the load caused/time elapsed by these commands (`device ls-by-daemon`)? Could they block/hang or cause any impact to the underlying devices?

I say this because most of the metrics so far came from ceph-mgr in-memory cache, but as long as we're adding more and more metrics, ceph-dashboard read-only operations are no longer 'harmless'.

Reply:
I unfortunately don't know things for sure, but I imagine that a failing drive may hang when SMART data is requested, although this would be somewhat odd, as this data is required to diagnose the health of the device.

But I'm also wary when collecting data and hence decided not to stick with the 10 seconds interval the data is gathered on the OSD page, but only request SMART data once an OSD is selected by the user. 

If we decide to add the history of reports, though, the rest of the data will come from a cache.

You might want to know that the commands used by this implementation are also used to periodically collect smart data for the `diskprediction` module. IIRC, it is recommended (or the default) to gather the data once a day. If there was an impact on performance or IO, this job would cause it, too.

Reply:
Thanks for the analysis. I was wondering if it makes sense to extend the ceph-mgr cache to other MonCommands.